text/2022-01-12-rpc-through-unix-socket.md

```diff
@@ -0,0 +1,242 @@
+
+This problem can be solved by opening an extra socket or pipe to communicate
+with a subprocess. As it's hard to move a named socket to another mount
+namespace, the abstract unix socket is the best choice.
+
+## Detailed Design
+
+When it comes to the communicating between chaos-daemon and its subprocess, we
+have three components need to describe:
+
+1. Chaos Daemon
+2. nsexec
+3. The subprocess
+
+### Chaos Daemon
+
+The bpm should enable the Chaos Daemon to pass an extra file (which is the unix
+socket fd) to the subprocess.
+
+```go
+func (b *ProcessBuilder) WithSocket() *ProcessBuilder {
+    b.WithSocket = true
+    return b
+}
+```
+
+While building this process, this field should be set to the `ExtraFiles`:
+
+```go
+// Build builds the process
+func (b *ProcessBuilder) Build() *ManagedProcess {
+    ...
+    if b.WithSocket {
+        rawListener, err := net.Listen("unix", fmt.Sprintf("@chaos-daemon-%s", *b.identifier))
+        listener := rawListener.(*net.UnixListener)
+        listenSocket, err := listener.File()
+        command.ExtraFiles = append(command.ExtraFiles, listenSocket)
+    }
+    ...
+}
+```
+
+Then the chaos daemon can send command by dialing this abstract socket:
+
+```go
+httpc := http.Client{
+    Transport: &http.Transport{
+        DialContext: func(_ context.Context, _, _ string) (net.Conn, error) {
+            return net.Dial("unix", fmt.Sprintf("@chaos-daemon-%s", *b.identifier))
+        },
+    },
+}
+```
+
+The name of the abstract unix socket is `@chaos-daemon-{b.identifier}`, which
+means communicating is only avaiable for process with an identifier. I think
+it's acceptable. As as alternative, we can generate an UUID as the name for
+every process, but I prefer the identifier (for the convenience of debug)
+
+ **Make sure that every LISTENING sockets are closed after the command has
+started, or the parent will be blocked by dialing dead children**.
+
+If every listening fd is closed, further request will get an error: `dial unix
+@xxxx: connect: connection refused`.
+
+### nsexec
+
+The newest version of `nsexec` already supports passing files to its subprocess,
+with the help of [command-fds](https://github.com/google/command-fds)
+
+### Subprocess
+
+The subprocess, for example `toda`, will need to establish its own transport
+from a raw fd. In go, the raw fd can be converted to a `os.File` directly:
+
+```go
+s := os.NewFile(3, "socket")
+
+listener, err := net.FileListener(s)
+```
+
+## Alternative
+
+1. Don't touch it!
+
+   Yes. This RFC doesn't provide any obvious improvement (except removing the
+   blocking buffer inside bpm), but I think it's valuable enough considering the
+   complexity of the blocking buffer and the further progress on rpc through
+   HTTP or gRPC... (But I agree that this proposal doesn't have high priority.)
+
+2. Dial abstract unix socket directly (rather than passing fd).
+
+   Unfortunately, the abstract unix socket is binded with the network namespace,
```

Comment:
I prefer this & actually we have access to mnt namespace of pod in chaos-daemon on proc file system.